actors/builtin/market/market_actor.go

```diff
@@ -427,78 +411,139 @@ func (a Actor) OnMinerSectorsTerminate(rt Runtime, params *OnMinerSectorsTermina
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to load deal opts set: %s", err)
+		}
 
-	// TODO: award some small portion of slashed to caller as incentive
+		updatesNeeded := make(map[abi.ChainEpoch][]abi.DealID)
 
-	_, code := rt.Send(builtin.BurntFundsActorAddr, builtin.MethodSend, nil, slashed)
-	builtin.RequireSuccess(rt, code, "failed to burn funds")
-	return nil
-}
+		states, err := AsDealStateArray(adt.AsStore(rt), st.States)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "get state state: %v", err)
+		}
 
-type HandleInitTimeoutDealsParams struct {
-	Deals []abi.DealID // TODO: RLE
-}
+		et, err := adt.AsBalanceTable(adt.AsStore(rt), st.EscrowTable)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "loading escrow table: %s", err)
+		}
 
-func (a Actor) HandleInitTimeoutDeals(rt Runtime, params *HandleInitTimeoutDealsParams) *adt.EmptyValue {
-	rt.ValidateImmediateCallerType(builtin.CallerTypesSignable...)
-	var st State
-	var verifiedDeals []*DealProposal
-	slashedAmount := rt.State().Transaction(&st, func() interface{} {
-		slashed := abi.NewTokenAmount(0)
-		for _, dealID := range params.Deals {
-			deal := st.mustGetDeal(rt, dealID)
-			state := st.mustGetDealState(rt, dealID)
-
-			// Deal has not been activated.
-			if state.SectorStartEpoch == epochUndefined {
-				// Now is after StartEpoch when the Deal should have been activated, hence clean up.
-				if rt.CurrEpoch() > deal.StartEpoch {
-					// Store VerifiedDeal to restore bytes for VerifiedClient.
-					if deal.VerifiedDeal {
-						verifiedDeals = append(verifiedDeals, deal)
+		lt, err := adt.AsBalanceTable(adt.AsStore(rt), st.LockedTable)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "loading locked balance table: %s", err)
+		}
+
+		for i := st.LastCron; i <= rt.CurrEpoch(); i++ {
+			if err := dbe.ForEach(i, func(dealID abi.DealID) error {
+				state, found, err := states.Get(dealID)
+				if err != nil {
+					rt.Abortf(exitcode.ErrIllegalState, "failed to get deal: %d", dealID)
+				}
+
+				if !found {
+					return nil
+				}
+
+				deal := st.mustGetDeal(rt, dealID)
+
+				if state.SectorStartEpoch == epochUndefined {
+					// Not yet appeared in proven sector; check for timeout.
+					if rt.CurrEpoch() > deal.StartEpoch {
+						slashed := st.processDealInitTimedOut(rt, et, lt, dealID, deal, state)
+						if !slashed.IsZero() {
+							amountSlashed = big.Add(amountSlashed, slashed)
+						}
+						if deal.VerifiedDeal {
+							timedOutVerifiedDeals = append(timedOutVerifiedDeals, deal)
+						}
+						return nil
 					}
-					newlySlashed := st.processDealInitTimedOut(rt, dealID)
-					big.Add(slashed, newlySlashed)
-				} else {
-					// All deals must have timed out.
-					rt.Abortf(exitcode.ErrIllegalArgument, "not all deals have timed out: %d", dealID)
+
+					// This should not be able to happen
+					rt.Abortf(exitcode.ErrIllegalState, "invalid deal state, unstarted, not timed out")
```

Comment:
I agree it shouldn't, but handling it is tricky. This abort will mean state is rolled back and on the next epoch the cron actor will call again with the same result, and never progress past here. Would it be better to `Assert` and halt execution entirely? Alternatively, don't abort here but let all subsequent processing happen (and like, that other place recently, we should look into logging facilities for the runtime). Or re-schedule it for the StartEpoch, thus failing toward correct behaviour.

Reply:
I'm leaning towards an assert for now, we can leave a TODO to look into it again later. Hopefully it never shows up